### ISSUE TITLE
Convex crashing silently

### ISSUE THREAD

**Issue by lpawela** [JuliaOpt/Convex.jl#282]
This is an extended restatement of the following issue https://github.com/ZKSI/QuantumInformation.jl/issues/50
I couldn't find any reason for the crashes on our side. Consider the following code
```julia
using Convex, SCS, LinearAlgebra

⊗ = kron
function norm_diamond_primal(J::Matrix, d1::Int, d2::Int, eps)
    
    X = ComplexVariable(d1*d2, d1*d2)
    t = 0.5*inner_product(X, J) + 0.5*inner_product(X', J')

    ρ₀ = ComplexVariable(d1, d1)
    ρ₁ = ComplexVariable(d1, d1)

    constraints = [ρ₀ in :SDP, ρ₁ in :SDP]
    constraints += tr(ρ₀) == 1
    constraints += tr(ρ₁) == 1
    iden = Matrix{ComplexF64}(I, d2, d2)
    constraints += [iden ⊗ ρ₀ X; X' iden ⊗ ρ₁] in :SDP

    problem = maximize(t, constraints)
    solve!(problem, SCSSolver(verbose=0, eps=eps))
    problem.optval
end

function ket(::Type{T}, val::Int, dim::Int) where T<:Number
    dim > 0 ? () : throw(ArgumentError("Vector dimension has to be nonnegative"))
    1 <= val <= dim ? () : throw(ArgumentError("Label have to be smaller than vector dimension"))
    ψ = zeros(T, dim)
    ψ[val] = one(T)
    ψ
end
ket(val::Int, dim::Int) = ket(ComplexF64, val, dim)
proj(ψ::AbstractVector{<:Number}) = ψ * ψ'


effs = Vector{ComplexF64}[[1/sqrt(2), 0], [1/sqrt(6), 1/sqrt(3)], [(-sqrt(2)-1im*sqrt(6))/(4sqrt(3)), 1/sqrt(3)], [(-sqrt(2)+1im*sqrt(6))/(4sqrt(3)), 1/sqrt(3)]]
effsperm = Vector{ComplexF64}[effs[2], effs[3], effs[4], effs[1]]
J1 = sum(proj(ket(i, 4)) ⊗ proj(ket(j, 4)) ⊗ proj(effs[i]) ⊗ proj(effs[j]) for i=1:4, j=1:4)
J2 = sum(proj(ket(i, 4)) ⊗ proj(ket(j, 4)) ⊗ proj(effsperm[i]) ⊗ proj(effsperm[j]) for i=1:4, j=1:4)

norm_diamond_primal(J1-J2, 4, 16, 1e-7)
```
This code crashes silently, meaning no exception thrown, yet there are no julia processes running. Killing it with `Ctrl+c` results in 
```
^C
signal (2): Interrupt
in expression starting at /home/lpawela/Untitled-1.jl:39
unknown function (ip: 0x7f3c21b216a9)
_mid_memalign at /build/glibc-OTsEL5/glibc-2.27/malloc/malloc.c:3312 [inlined]
__posix_memalign at /build/glibc-OTsEL5/glibc-2.27/malloc/malloc.c:5369
jl_malloc_aligned at /buildworker/worker/package_linux64/build/src/julia_internal.h:869 [inlined]
jl_gc_big_alloc at /buildworker/worker/package_linux64/build/src/gc.c:872
jl_gc_alloc_ at /buildworker/worker/package_linux64/build/src/julia_internal.h:268 [inlined]
jl_gc_alloc at /buildworker/worker/package_linux64/build/src/gc.c:2911
jl_gc_alloc_buf at /buildworker/worker/package_linux64/build/src/julia_internal.h:290 [inlined]
jl_reserve_excstack at /buildworker/worker/package_linux64/build/src/rtutils.c:315 [inlined]
jl_push_excstack at /buildworker/worker/package_linux64/build/src/rtutils.c:327
throw_internal at /buildworker/worker/package_linux64/build/src/task.c:388
jl_sig_throw at /buildworker/worker/package_linux64/build/src/task.c:439
unknown function (ip: 0xffffffffffffffff)
unknown function (ip: 0xffffffffffffffff)
Allocations: 37492817 (Pool: 37456441; Big: 36376); GC: 104
^C
signal (2): Interrupt
in expression starting at /home/lpawela/Untitled-1.jl:39
unknown function (ip: 0x7f3c21b216a9)
__libc_free at /build/glibc-OTsEL5/glibc-2.27/malloc/malloc.c:4266
uv__finish_close at /buildworker/worker/package_linux64/build/deps/srccache/libuv-2348256acf5759a544e5ca7935f638d2bc091d60/src/unix/core.c:279 [inlined]
uv__run_closing_handles at /buildworker/worker/package_linux64/build/deps/srccache/libuv-2348256acf5759a544e5ca7935f638d2bc091d60/src/unix/core.c:293 [inlined]
uv_run at /buildworker/worker/package_linux64/build/deps/srccache/libuv-2348256acf5759a544e5ca7935f638d2bc091d60/src/unix/core.c:363
jl_atexit_hook at /buildworker/worker/package_linux64/build/src/init.c:287
jl_exit at /buildworker/worker/package_linux64/build/src/jl_uv.c:592
jl_exit_thread0_cb at /buildworker/worker/package_linux64/build/src/signals-unix.c:344
unknown function (ip: 0xffffffffffffffff)
unknown function (ip: 0xffffffffffffffff)
Allocations: 37492817 (Pool: 37456441; Big: 36376); GC: 104
^C
signal (2): Interrupt
in expression starting at /home/lpawela/Untitled-1.jl:39
unknown function (ip: 0x7f3c21b216a9)
__libc_free at /build/glibc-OTsEL5/glibc-2.27/malloc/malloc.c:4266
__run_exit_handlers at /build/glibc-OTsEL5/glibc-2.27/stdlib/exit.c:108
exit at /build/glibc-OTsEL5/glibc-2.27/stdlib/exit.c:139
jl_exit_thread0_cb at /buildworker/worker/package_linux64/build/src/signals-unix.c:347
unknown function (ip: 0xffffffffffffffff)
unknown function (ip: 0xffffffffffffffff)
Allocations: 37492817 (Pool: 37456441; Big: 36376); GC: 104
```

**Comment by ararslan** [comment]
> yet there are no julia processes running

I'm not sure how that's possible given that the backtraces from the interrupts are coming from Julia's signal handler.

**Comment by lpawela** [comment]
Yes, sorry I made this not clear. There is no processor load from julia processes.

**Comment by ararslan** [comment]
Oh, gotcha. So it's just stuck running but doing nothing.

**Comment by lpawela** [comment]
Yes. In `htop` I see some load during the initialization, but when the code hits the last line all CPU load drops to zero.

**Comment by lpawela** [comment]
I noticed I referenced the wrong issue. It should be https://github.com/ZKSI/QuantumInformation.jl/issues/50. I updated the original post as well.

**Comment by ericphanson** [comment]
I tried running this code on the current release of Convex (v0.12.0), and got a `StackOverflowError`:
```julia
julia> include("282.jl")
[ Info: Recompiling stale cache file /Users/ephhome/.julia/compiled/v1.1/Convex/Sr0e5.ji for Convex [f65535da-76fb-5f13-bab9-19810c17039a]
ERROR: LoadError: StackOverflowError:
Stacktrace:
 [1] promote_typeof(::Int64, ::Int64, ::Vararg{Int64,N} where N) at ./promotion.jl:264 (repeats 25951 times)
 [2] vect(::Int64, ::Vararg{Int64,N} where N) at /Users/ephhome/.julia/packages/Convex/HwTSz/src/atoms/affine/stack.jl:139
 [3] #20 at ./none:0 [inlined]
 [4] iterate at ./generator.jl:47 [inlined]
 [5] collect_to!(::Array{Tuple{Symbol,Array{Int64,1}},1}, ::Base.Generator{Array{Tuple{Symbol,UnitRange{Int64}},1},getfield(SCS, Symbol("##20#22"))}, ::Int64, ::Int64) at ./array.jl:651
 [6] collect_to_with_first!(::Array{Tuple{Symbol,Array{Int64,1}},1}, ::Tuple{Symbol,Array{Int64,1}}, ::Base.Generator{Array{Tuple{Symbol,UnitRange{Int64}},1},getfield(SCS, Symbol("##20#22"))}, ::Int64) at ./array.jl:630
 [7] collect(::Base.Generator{Array{Tuple{Symbol,UnitRange{Int64}},1},getfield(SCS, Symbol("##20#22"))}) at ./array.jl:611
 [8] loadproblem!(::SCS.SCSMathProgModel, ::Array{Float64,1}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::Array{Float64,1}, ::Array{Tuple{Symbol,UnitRange{Int64}},1}, ::Array{Tuple{Symbol,UnitRange{Int64}},1}) at /Users/ephhome/.julia/packages/SCS/L1XbN/src/MPB_wrapper.jl:370
 [9] load_problem!(::SCS.SCSMathProgModel, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::Array{Tuple{Symbol,UnitRange{Int64}},1}, ::Array{Symbol,1}) at /Users/ephhome/.julia/packages/Convex/HwTSz/src/solution.jl:89
 [10] #solve!#4(::Bool, ::Bool, ::Bool, ::Function, ::Problem) at /Users/ephhome/.julia/packages/Convex/HwTSz/src/solution.jl:37
 [11] solve! at /Users/ephhome/.julia/packages/Convex/HwTSz/src/solution.jl:21 [inlined]
 [12] #solve!#3 at /Users/ephhome/.julia/packages/Convex/HwTSz/src/solution.jl:13 [inlined]
 [13] solve! at /Users/ephhome/.julia/packages/Convex/HwTSz/src/solution.jl:12 [inlined]
 [14] norm_diamond_primal(::Array{Complex{Float64},2}, ::Int64, ::Int64, ::Float64) at /Users/ephhome/Dropbox/Research/Code/julia-projects/ConvexTests/282/282.jl:19
 [15] top-level scope at none:0
 [16] include at ./boot.jl:326 [inlined]
 [17] include_relative(::Module, ::String) at ./loading.jl:1038
 [18] include(::Module, ::String) at ./sysimg.jl:29
 [19] include(::String) at ./client.jl:403
 [20] top-level scope at none:0
in expression starting at /Users/ephhome/Dropbox/Research/Code/julia-projects/ConvexTests/282/282.jl:39
```

It seems to be repeatedly calling `vect` for some reason, and in particular, Convex's reimplementation,
```
 Base.vect(args::Value...) = copyto!(Vector{Base.promote_typeof(args...)}(undef, length(args)), args)
```
Since I recently proposed removing Convex's method for `vect` (https://github.com/JuliaOpt/Convex.jl/pull/293), I tried the code on that branch as well. This led to high CPU usage for a couple minutes and then
```julia
┌ Warning: Problem status UnknownError; solution may be inaccurate.
└ @ Convex ~/Dropbox/LinkedFolders/Julia/dev/Convex/src/solution.jl:48
1.9999971978944495
```

I ran it again with `verbose=1` to see what SCS says (usually it gives more information than Convex's `UnknownError`) and saw it ran for the 5k iterations that it is allowed by default, but couldn't get to the desired accuracy in that time:

```julia
Status: Solved/Inaccurate
Hit max_iters, solution may be inaccurate
Timing: Solve time: 1.42e+02s
        Lin-sys: avg # CG iterations: 5.68, avg solve time: 2.15e-03s
        Cones: avg projection time: 1.38e-02s
        Acceleration: avg step time: 1.04e-02s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 1.4814e-06, dist(y, K*) = 1.9345e-05, s'y/|s||y| = 2.6321e-07
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 2.3093e-06
dual res:   |A'y + c|_2 / (1 + |c|_2) = 1.4581e-04
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 2.2329e-07
----------------------------------------------------------------------------
c'x = -2.0000, -b'y = -2.0000
============================================================================
```

I don't think that is a problem with Convex; my understanding is that SCS is good for solving large problems to low accuracy (e.g. `eps=1e-4`), but as a first order method is it not suited to choosing higher accuracies like `eps=1e-7`. SCS could successfully solve the problem (again on the branch https://github.com/JuliaOpt/Convex.jl/pull/293) with `eps=1e-4` in 3.5k iterations:
```julia
Status: Solved
Timing: Solve time: 8.88e+01s
        Lin-sys: avg # CG iterations: 5.80, avg solve time: 1.97e-03s
        Cones: avg projection time: 1.19e-02s
        Acceleration: avg step time: 9.58e-03s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 7.4014e-07, dist(y, K*) = 1.6959e-07, s'y/|s||y| = 4.4751e-07
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 1.4299e-06
dual res:   |A'y + c|_2 / (1 + |c|_2) = 7.8818e-05
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 2.3571e-07
----------------------------------------------------------------------------
c'x = -2.0000, -b'y = -2.0000
============================================================================
1.9999962490689098
```

**Comment by ararslan** [comment]
Fixed by #293. Please feel free to reopen if you run into it again.
